### PR TITLE
Add --print-bytes argument so that a subset of the packet can be printed.

### PR DESCRIPTION
**Pull request by stevekay**
Permit printing of a portion of the packet (minus link layer) when --print-bytes used with -X.

Example:

```
    $ ./tcpdump -h
    tcpdump version 4.8.0-PRE-GIT_2015_03_08
    libpcap version 1.8.0-PRE-GIT_2015_03_07
    OpenSSL 1.0.1e-fips 11 Feb 2013
    Usage: tcpdump [-aAbdDefhHIJKlLnNOpqRStuUvxX#] [ -B size ] [ -c count ]
                    [ -C file_size ] [ -E algo:secret ] [ -F file ] [ -G seconds ]
                    [ -i interface ] [ -j tstamptype ] [ -M secret ] [ --number ]
                    [ --print-bytes start:end ] [ -Q in|out|inout ]
                    [ -r file ] [ -s snaplen ] [ --time-stamp-precision precision ]
                    [ --immediate-mode ] [ -T type ] [ --version ] [ -V file ]
                    [ -w file ] [ -W filecount ] [ -y datalinktype ] [ -z command ]
                    [ -Z user ] [ expression ]
    $ ./tcpdump -r tests/mptcp.pcap -c 1 -XX
    reading from file tests/mptcp.pcap, link-type EN10MB (Ethernet)
    05:56:35.701161 IP 10.2.1.2.35961 > 10.1.1.2.ssh: Flags [S], seq 2912457561, win 14600, options         [mss 1460,sackOK,TS val 4294943152 ecr 0,nop,wscale 6,mptcp capable csum {0x9c9eabd1e46a33b2}],         length 0
            0x0000:  1651 5304 3f55 f28c f524 1b21 0800 4500  .QS.?U...$.!..E.
            0x0010:  0048 32e9 4000 4006 f1c0 0a02 0102 0a01  .H2.@.@.........
            0x0020:  0102 8c79 0016 ad98 9359 0000 0000 d002  ...y.....Y......
            0x0030:  3908 da99 0000 0204 05b4 0402 080a ffff  9...............
            0x0040:  a1b0 0000 0000 0103 0306 1e0c 0081 9c9e  ................
            0x0050:  abd1 e46a 33b2                           ...j3.
    $ ./tcpdump -r tests/mptcp.pcap -c 1 -X
    reading from file tests/mptcp.pcap, link-type EN10MB (Ethernet)
    05:56:35.701161 IP 10.2.1.2.35961 > 10.1.1.2.ssh: Flags [S], seq 2912457561, win 14600, options [mss 1460,sackOK,TS val 4294943152 ecr 0,nop,wscale 6,mptcp capable csum {0x9c9eabd1e46a33b2}], length 0
            0x0000:  4500 0048 32e9 4000 4006 f1c0 0a02 0102  E..H2.@.@.......
            0x0010:  0a01 0102 8c79 0016 ad98 9359 0000 0000  .....y.....Y....
            0x0020:  d002 3908 da99 0000 0204 05b4 0402 080a  ..9.............
            0x0030:  ffff a1b0 0000 0000 0103 0306 1e0c 0081  ................
            0x0040:  9c9e abd1 e46a 33b2                      .....j3.
    $ ./tcpdump -r tests/mptcp.pcap -c 1 -X --print-bytes 16:31
    reading from file tests/mptcp.pcap, link-type EN10MB (Ethernet)
    05:56:35.701161 IP 10.2.1.2.35961 > 10.1.1.2.ssh: Flags [S], seq 2912457561, win 14600, options [mss 1460,sackOK,TS val 4294943152 ecr 0,nop,wscale 6,mptcp capable csum {0x9c9eabd1e46a33b2}], length 0
            0x0010:  0a01 0102 8c79 0016 ad98 9359 0000 00    .....y.....Y....
    $ 
```

### CODE REVIEW

**Comment by mcr** [comment]
Please rebase, and please add one or two test cases that shows how this is used with an example packet.

File: interface.h [hunk 1]
```diff
@@ -224,6 +224,9 @@ extern netdissect_options *gndo;
 #define snaplen     gndo->ndo_snaplen
 #define snapend     gndo->ndo_snapend
 
+#define printstart gndo->ndo_snaplen
+#define printend gndo->ndo_snaplen
```
Comment (mcr) [inline interface.h]: If possible, please change the code to use ndo-> explicitely, rather than introducing new macros.

File: tcpdump.c [hunk 2]
```diff
@@ -1273,6 +1277,21 @@ main(int argc, char **argv)
 			++pflag;
 			break;
 
+		case OPTION_PRINT_BYTES:
+			if ((colon=strchr(optarg,':')) == 0) {
```
Comment (mcr) [inline tcpdump.c]: maybe { char *colon; ...} so that we don't need another function-wide automatic.

File: tcpdump.c [hunk 3]
```diff
@@ -2480,8 +2499,13 @@ print_packet(u_char *user, const struct pcap_pkthdr *h, const u_char *sp)
 			 * print nothing.
 			 */
 			if (h->caplen > hdrlen)
-				hex_and_ascii_print(ndo, "\n\t", sp + hdrlen,
-				    h->caplen - hdrlen);
+				if(ndo->ndo_print_bytes_flag) {
+					hex_and_ascii_print_with_offset(ndo, "\n\t", sp + hdrlen + ndo->ndo_print_start,
+				    	ndo->ndo_print_end-ndo->ndo_print_start+1,ndo->ndo_print_start);
+				} else {
+					hex_and_ascii_print(ndo, "\n\t", sp + hdrlen,
```
Comment (mcr) [inline tcpdump.c]: could we simplify and just call _with_offset() with 0..max?

File: interface.h (hunk 1 above)
Comment (guyharris) [reply interface.h]: Especially given that there *IS* no `gndo` variable any more.

**Comment by infrastation** [comment]
This is the longest-standing pull request in this project. Unfortunately, nobody had time to review it properly in 2015, perhaps in part because the use case for this feature still isn't clear. The review was done more than 1.5 years ago. Unless someone is willing to work on these changes (including the original author, @stevekay), this pull request will be closed in 7 days.

**Comment by infrastation** [comment]
It is time.

**Pull request closed without merging**